### PR TITLE
listener: fix ipv6 error

### LINKED ISSUE #3911: Error Creating Listeners on Machines that does not support IPv6 Family
https://github.com/envoyproxy/envoy/commit/01d2e16e973a76cc05f5e9e5c7a74f73498bc6c2

*Description*:
After the PR https://github.com/envoyproxy/envoy/pull/3851 (commit https://github.com/envoyproxy/envoy/commit/01d2e16e973a76cc05f5e9e5c7a74f73498bc6c2) we are not able to create Listeners from ADS on machines where IPv6 family is not enabled.

The error we get is 
`07:09:56 2018/07/19 14:09:56 Envoy STDERR: [2018-07-19 14:09:56.284][26][warning][upstream] source/common/config/grpc_mux_impl.cc:223] gRPC config for type.googleapis.com/envoy.api.v2.Listener update rejected: Error adding/updating listener testListener: IPv6 addresses are not supported on this machine: [::]:0`

*Repro steps*:
Try to create Listener with ADS(any xDS) with the following Listener configuration. There is nothing special in the Listener configuration. Also please note that we are not using the new `FilterMatch`.
```
listener := v2.Listener{
		Name: "testListener",
		Address: core.Address{
			Address: &core.Address_SocketAddress{
				SocketAddress: &core.SocketAddress{
					Protocol: core.TCP,
					Address:  localhost,
					PortSpecifier: &core.SocketAddress_PortValue{
						PortValue: uint32(6666),
					},
				},
			},
		},
		FilterChains: []listener.FilterChain{{
			Filters: []listener.Filter{{
				Name:   util.HTTPConnectionManager,
				Config: managerStruct,
			}},
		}},
	}

```

### PR DESCRIPTION
**Pull request by ramaraochavali**
Signed-off-by: Rama <rama.rao@salesforce.com>
*Description*:
Fixes the bug https://github.com/envoyproxy/envoy/issues/3911
*Risk Level*: Low
*Testing*: Automated
*Docs Changes*: N/A
*Release Notes*: N/A

Fixes #3911

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@PiotrSikora I am not 100% sure if this is the right fix because I am fully familiar with this part. PTAL and suggest. I can change based on your feedback.

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -380,11 +380,18 @@ void ListenerImpl::convertDestinationIPsMapToTrie() {
...
+              std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
+                  std::make_shared<ServerNamesMap>(entry.second),
+                  {Network::Address::CidrRange::create("0.0.0.0/0"),
+                   Network::Address::CidrRange::create("::/0")}));
+        } else {
+          list.push_back(
+              std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
+                  std::make_shared<ServerNamesMap>(entry.second),
+                  {Network::Address::CidrRange::create("0.0.0.0/0")}));
+        }
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.cc]: I think we need support IPv6-only setups as well:
```
diff --git a/source/server/listener_manager_impl.cc b/source/server/listener_manager_impl.cc
index 393fdfe1e..e69bebe4b 100644
--- a/source/server/listener_manager_impl.cc
+++ b/source/server/listener_manager_impl.cc
@@ -380,17 +380,19 @@ void ListenerImpl::convertDestinationIPsMapToTrie() {
     for (const auto& entry : destination_ips_map) {
       std::vector<Network::Address::CidrRange> subnets;
       if (entry.first == EMPTY_STRING) {
-        list.push_back(
-            std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
-                std::make_shared<ServerNamesMap>(entry.second),
-                {Network::Address::CidrRange::create("0.0.0.0/0"),
-                 Network::Address::CidrRange::create("::/0")}));
+        if (Network::Address::ipFamilySupported(AF_INET)) {
+          subnets.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
+        }
+        if (Network::Address::ipFamilySupported(AF_INET6)) {
+          subnets.push_back(Network::Address::CidrRange::create("::/0"));
+        }
       } else {
-        list.push_back(
-            std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
-                std::make_shared<ServerNamesMap>(entry.second),
-                {Network::Address::CidrRange::create(entry.first)}));
+        subnets.push_back(Network::Address::CidrRange::create(entry.first));
       }
+      list.push_back(
+          std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
+              std::make_shared<ServerNamesMap>(entry.second),
+              std::vector<Network::Address::CidrRange>(subnets)));
     }
     destination_ips_pair.second = std::make_unique<DestinationIPsTrie>(list, true);
   }
```

Comment (PiotrSikora) [reply source/server/listener_manager_impl.cc]: cc @ggreenway

Comment (ggreenway) [reply source/server/listener_manager_impl.cc]: I agree with @PiotrSikora

File: source/server/listener_manager_impl.cc [hunk 2]
```diff
@@ -380,11 +380,14 @@ void ListenerImpl::convertDestinationIPsMapToTrie() {
...
-                {Network::Address::CidrRange::create("0.0.0.0/0"),
-                 Network::Address::CidrRange::create("::/0")}));
+        std::vector<Network::Address::CidrRange> cidr_ranges;
+        if (Network::Address::ipFamilySupported(AF_INET)) {
+          cidr_ranges.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
+        }
+        if (Network::Address::ipFamilySupported(AF_INET6)) {
+          cidr_ranges.push_back(Network::Address::CidrRange::create("::/0"));
+        }
+        list.push_back({std::make_shared<ServerNamesMap>(entry.second), cidr_ranges});
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.cc]: Could you use the patch I provided (using unused `subnets` and sharing more of the code with the `!= EMPTY_STRING` path)? It's a bit more readable, IMHO.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: Updated. Please check

**Comment by ramaraochavali** [comment]
@ggreenway Thanks. I was thinking about such a test and checked to see if we can write a mock but looks like for functions like this writing mock is not possible with gmocks. If there is any possible way to do it, do let me know. I will look in to it.

**Comment by ggreenway** [comment]
@ramaraochavali The way to mock this is to add socket() to `OsSysCalls` (see envoy/include/envoy/api/os_sys_calls.h). Then use `OsSysCalls::socket()` in `ipFamilySupported()`. You can then mock the behavior of `::socket()`.

**Pull request merged**